src/optimizer/remove_unused_columns.cpp

```diff
@@ -287,6 +374,104 @@ void RemoveUnusedColumns::VisitOperator(LogicalOperator &op) {
 	}
 }
 
+static idx_t GetColumnIdsIndexForFilter(vector<ColumnIndex> &column_ids, idx_t filter_idx) {
+	// Find the index in the column_ids that contains the column referenced by the filter
+	auto it = std::find_if(column_ids.begin(), column_ids.end(), [&filter_idx](const ColumnIndex &column_index) {
+		return column_index.GetPrimaryIndex() == filter_idx;
+	});
+	if (it == column_ids.end()) {
+		throw InternalException("Could not find column index for table filter");
+	}
+	return static_cast<idx_t>(std::distance(column_ids.begin(), it));
+}
+
+static ColumnIndex PathToIndex(const vector<idx_t> &path) {
+	D_ASSERT(!path.empty());
+	ColumnIndex index = ColumnIndex(path[0]);
+	reference<ColumnIndex> current(index);
+	for (idx_t i = 1; i < path.size(); i++) {
+		current.get().AddChildIndex(ColumnIndex(path[i]));
+		current = current.get().GetChildIndex(0);
+	}
+	return index;
+}
+
+static void WritePushdownExtractColumns(ReferencedColumn &col, const LogicalType &column_type,
+                                        BindingsRewriteState &state, idx_t col_sel_idx) {
+	auto &old_column_ids = state.old_column_ids;
+	auto struct_column_index = old_column_ids[col_sel_idx].GetPrimaryIndex();
+
+	//! For each struct extract, replace the expression with a BoundColumnRefExpression
+	//! The expression references a binding created for the extracted path, 1 per unique path
+	for (auto &struct_extract : col.struct_extracts) {
+		//! Replace the top-level struct extract with a BoundColumnRefExpression
+		//! FIXME: We can only replace the top-level struct if the path is in 'col.unique_paths'
+		//! Otherwise there is a parent path that's referenced, so we don't push the extract into the storage all the
+		//! way
+
+		auto &full_path = struct_extract.extract_path;
+		D_ASSERT(!full_path.empty());
+		idx_t depth = 0;
+		auto end_iterator = ++full_path.begin();
+		auto entry = col.unique_paths.end();
+		while (true) {
+			vector<idx_t> copy(full_path.begin(), end_iterator);
```

Comment:
we can do this more efficiently, we can just keep adding to the vector, no need to fully rewrite it every iteration